src/ZVal.cc

```diff
@@ -0,0 +1,354 @@
+		uint_val = vu.uint_val;
+		break;
+
+	case zeek::TYPE_DOUBLE:
+	case zeek::TYPE_INTERVAL:
+	case zeek::TYPE_TIME:
+		double_val = vu.double_val;
+		break;
+
+	case zeek::TYPE_FUNC:
+		func_val = vu.func_val;
+		Ref(func_val);
+		break;
+
+	case zeek::TYPE_FILE:
+		file_val = vu.file_val;
+		Ref(file_val);
+		break;
+
+	case zeek::TYPE_LIST:
+		list_val = v.release()->AsListVal();
+		break;
+
+	case zeek::TYPE_OPAQUE:
+		opaque_val = v.release()->AsOpaqueVal();
+		break;
+
+	case zeek::TYPE_PATTERN:
+		re_val = v.release()->AsPatternVal();
+		break;
+
+	case zeek::TYPE_TABLE:
+		table_val = v.release()->AsTableVal();
+		break;
+
+	case zeek::TYPE_VECTOR:
+		{
+		vector_val = v.release()->AsVectorVal();
+
+		// Some run-time type-checking, sigh.
+		auto my_ytag = t->AsVectorType()->Yield()->Tag();
+		auto v_ytag = vt->AsVectorType()->Yield()->Tag();
+
+		if ( my_ytag != v_ytag && my_ytag != zeek::TYPE_ANY &&
+		     v_ytag != zeek::TYPE_ANY )
+			{
+			// Despite the above checks, this clash can still
+			// happen thanks to the intercession of vector-of-any,
+			// which for example can allow a function to return
+			// a concrete vector-of-X that's assigned to a local
+			// with a concrete vector-of-Y type.
+			reporter->Error("vector type clash: %s vs. %s (%s)",
+					type_name(my_ytag), type_name(v_ytag),
+					obj_desc(v));
+			if ( zval_error_addr )
+				*zval_error_addr = true;
+			}
+
+		break;
+		}
+
+	case zeek::TYPE_RECORD:
+		record_val = v.release()->AsRecordVal();
+		break;
+
+	case zeek::TYPE_STRING:
+		string_val = v.release()->AsStringVal();
+		break;
+
+	case zeek::TYPE_ADDR:
+		addr_val = v.release()->AsAddrVal();
+		break;
+
+	case zeek::TYPE_SUBNET:
+		subnet_val = v.release()->AsSubNetVal();
+		break;
+
+	case zeek::TYPE_ANY:
+		any_val = v.release();
+		break;
+
+	case zeek::TYPE_TYPE:
+		type_val = t->Ref();
+		break;
+
+	case zeek::TYPE_ERROR:
+	case zeek::TYPE_TIMER:
+	case zeek::TYPE_UNION:
+	case zeek::TYPE_VOID:
+		reporter->InternalError("bad type in ZAMValUnion constructor");
+	}
+	}
+
+bool ZAMValUnion::IsNil(const TypePtr& t) const
+	{
+	switch ( t->Tag() ) {
+	case zeek::TYPE_ADDR:		return ! addr_val;
+	case zeek::TYPE_ANY:		return ! any_val;
+	case zeek::TYPE_FILE:		return ! file_val;
+	case zeek::TYPE_FUNC:		return ! func_val;
+	case zeek::TYPE_LIST:		return ! list_val;
+	case zeek::TYPE_OPAQUE:	return ! opaque_val;
+	case zeek::TYPE_PATTERN:	return ! re_val;
+	case zeek::TYPE_RECORD:	return ! record_val;
+	case zeek::TYPE_STRING:	return ! string_val;
+	case zeek::TYPE_SUBNET:	return ! subnet_val;
+	case zeek::TYPE_TABLE:	return ! table_val;
+	case zeek::TYPE_TYPE:		return ! type_val;
+
+	default:	return false;
+	}
+	}
+
+ValPtr ZAMValUnion::ToVal(const TypePtr& t) const
+	{
+	Val* v;
+
+	switch ( t->Tag() ) {
+	case zeek::TYPE_INT:
+		// We can't use make_intrusive directly because this
+		// constructor is protected, sigh.
+		v = new Val(int_val, zeek::TYPE_INT);
+		return {AdoptRef{}, v};
```

Comment:
There's `val_mgr->Int()` (and also `val_mgr->Bool()` / `val_mgr->Count()` for bool and count types below) that would give back a `ValPtr` to a preallocated-or-else-newly-constructed object.

Reply:
Thanks for flagging, my next push will include this.